lib/virtio/virtqueue.c

```diff
@@ -8,22 +8,34 @@
 #include <string.h>
 #include <openamp/virtqueue.h>
 #include <metal/atomic.h>
-#include <metal/dma.h>
-#include <metal/io.h>
 #include <metal/log.h>
 #include <metal/alloc.h>
 
 /* Prototype for internal functions. */
 static void vq_ring_init(struct virtqueue *, void *, int);
 static void vq_ring_update_avail(struct virtqueue *, uint16_t);
 static uint16_t vq_ring_add_buffer(struct virtqueue *, struct vring_desc *,
-				   uint16_t, struct metal_sg *, int, int);
+				   uint16_t, struct virtqueue_buf *, int, int);
 static int vq_ring_enable_interrupt(struct virtqueue *, uint16_t);
 static void vq_ring_free_chain(struct virtqueue *, uint16_t);
 static int vq_ring_must_notify_host(struct virtqueue *vq);
 static void vq_ring_notify_host(struct virtqueue *vq);
 static int virtqueue_nused(struct virtqueue *vq);
 
+/* Default implementation of P2V based on libmetal */
```

Comment:
> why not us week implementation in metal_io_phys_to_virt? for me this is the role of libmetal...
> do you think that will save code to do it here?

The reason is that metal_io_phys_to_virt works on a metal_io_region basis, so a given system could reasonable be using metal_io_phys_to_virt for other purposes outside of openamp.  So we could possibly optimize if we explicitly know if there is only a single metal_io_region in the system, but once there are multiple you cant.

However, doing this on virtqueue side means we know there is only a single region in play specifically for the virtqueue and thus can optimize with that knowledge in place.

Reply:
my concern  is that application can also want to access to this region based on a virtual address ( seems not the case today but for DMA transfer, could be the case).  
if these function are overwritten,  application should call  this function instead of metal_io_phys_to_virt to  ensure coherency in memory access translations.

Reply:
> my concern is that application can also want to access to this region based on a virtual address ( seems not the case today but for DMA transfer, could be the case).
> if these function are overwritten, application should call this function instead of metal_io_phys_to_virt to ensure coherency in memory access translations.

* the user can always use the default metal_io_region based solution if they want this functionality.
* Do you really expect P2V or V2P to handle coherency?
* if the app/system provides its own version, than I expect it handles/takes ownership of any memory mgmt issues like coherency.

Reply:
My concern is if we allow to have this override, that is we open a hole for user not to use metal I/O, however, we assumes that in OpenAMP, we use libmetal for the memory and I/O abstraction. but in this case, we cannot make this assumption, and then the library can have potential issues.